Created INavigationManager to replace Globals.NavigateURL

ahoefling opened the pull request:
Closes: #3159

## Summary
Refactored `Globals.NavigateURL()` to a new interface `INavigationManager` and updated **ALL** references to use dependency injection.

### Note
This is a first attempt, and we may need to make several revisions to this or target 10.0 instead of 9.x.

## Tasks
Below is a list of tasks that I need to complete prior to us accepting the PR. I will continue to edit them as we review the code changes:

- [x] Fix failing unit tests
- [x] Verify installer works
- [x] Verify DNN Happy Path - General workflow, login, logout, page management, etc.
- [x] Create `DotNetNuke.Abstractions` library to contain new common registrations such as `INavigationManager`
- [x] Create nuspec file for `DotNetNuke.Abstractions` so it can be registered as a NuGet library
- [x] Update all `.ascx.cs` injections to be `private readonly` instead of `protected`
- [x] Remove extra xml records added by Visual Studio (.csproj, .config files)
- [x] Add new unit tests for `DotNetNuke.Common.NavigationManager`

# All Tasks Completed

ahoefling commented on DNN Platform/Library/Properties/AssemblyInfo.cs:
```diff
@@ -56,4 +56,14 @@
+[assembly: InternalsVisibleTo("DotNetNuke.Modules.Groups")] // Once Globals is refeactored to Dependency Injection we should be able to remove this
+[assembly: InternalsVisibleTo("DotNetNuke.Modules.Journal")] // Once Globals is refeactored to Dependency Injection we should be able to remove this
+[assembly: InternalsVisibleTo("DotNetNuke.Modules.RazorHost")] // Once Globals is refeactored to Dependency Injection we should be able to remove this
+[assembly: InternalsVisibleTo("DotNetNuke.Website")] // Once Globals is refeactored to Dependency Injection we should be able to remove this
```
The `Globals.DependencyProvider` (our DependencyInjection Container) is internal by design. However the entire platform doesn't fully support Dependency Injection and there are many areas where we need to manually resolve dependencies using the Service Locator Pattern

```
IMyCutomType myInstance = Globals.DependencyProvider.GetService<IMyCustomType>();
```

I am not sure the best way to approach this problem as to make this change I need all of those `InternalsVisibleTo` declarations otherwise we won't have access to the `DependencyProvider`.

@mitchelsellers and I have discussed in the past that we do not want to make the `DependencyProvider` public and it is internal to the platform.

bdukes replied:
Gotcha, that works for me.

ahoefling commented on DNN Platform/Library/DotNetNuke.Library.csproj:
```diff
@@ -189,6 +189,8 @@
     <Compile Include="Application\AssemblyStatusAttribute.cs" />
     <Compile Include="Application\DotNetNukeContext.cs" />
+    <Compile Include="Common\Interfaces\INavigationManager.cs" />
+    <Compile Include="Common\NavigationManager.cs" />
```
I really like this idea!

I'll update the PR to include a new project called `DotNetNuke.Common.Interfaces` which is where we can place any of our shared interfaces such as `INavigationManager`

bdukes replied:
Some of the Microsoft packages make me think we might want a name like `DotNetNuke.Abstractions`.  Does that sound right to you?  Or would you understand that to be a different concept?

ahoefling replied:
I didn't even think of that, I love it!

ahoefling replied:
I think I have this figured out and will push changes soon. However, this creates a very large rabbit hole of dependencies due to the `PortalSettings` usage which now creates a cicrular dependency between the new `DotNetNuke.Abstractions` and `DotNetNuke.Library`.

I'll document areas of concern with the new code push

ahoefling commented on DNN Platform/Library/Common/Interfaces/INavigationManager.cs:
```diff
@@ -0,0 +1,116 @@
+        /// Gets the URL to the current page.
+        /// </summary>
+        /// <returns>Formatted URL.</returns>
+        [Browsable(false), DesignerSerializationVisibility(DesignerSerializationVisibility.Hidden)]
```
I just copied this over without really researching the initial intent, I'm all for removing these attributes

ahoefling replied:
These have been removed as requested

ahoefling commented on DNN Platform/Library/Common/Globals.cs:
```diff
@@ -2898,10 +2901,10 @@ public static string UserProfileURL(int userId)
         {
-            PortalSettings portalSettings = PortalController.Instance.GetCurrentPortalSettings();
-            return NavigateURL(portalSettings.ActiveTab.TabID, Null.NullString);
+            return DependencyProvider.GetService<INavigationManager>().NavigateURL();
```
This makes sense to me, we can update this to use `GetRequiredService` which will throw an exception if it can't resolve `INavigationManager`.

The less chance of a `NullReferenceException` the better

bdukes replied:
I've pushed b669f8dc608acbfc19c640cf4eff0bdf8f38d26e with this change

ahoefling commented on DNN Platform/Admin Modules/Dnn.Modules.Console/ViewConsole.ascx.cs:
```diff
@@ -53,13 +55,19 @@ namespace Dnn.Modules.Console
 {
 	public partial class ViewConsole : PortalModuleBase
 	{
+        protected INavigationManager NavigationManager { get; }
```
This is a common standard that I have followed in many of my projects which allows any child class to use the injected type. In the `UserControl` scenario I agree with you and we can update all of these implementation to be `private readonly`

bdukes replied:
Sure, if the class isn't `sealed`, leaving them `protected` as a general practice makes sense to me.  Does it make sense to move a commonly used dependency like `INavigationManager` into `PortalModuleBase`, or would you prefer each implementation have to ask for it?

ahoefling replied:
This is a practice that DNN has done historically, created protected properties in parent classes making it easy to access. A practice that I would like to get away from as it now creates a tight-coupling between `PortalModuleBase` and `INavigationManager` (you can't have one without the other).

We should require all child classes to request it if needed. An exception to this rule is if **EVERY** child class uses it then we could put it in the parent class, but I don't believe every class will be using it, just a lot of classes use it

ahoefling commented:
@bdukes I updated the code with a notable change of migrating `INavigationManager` to a new .NET Standard 2.0 project called `DotNetNuke.Abstractions`. A small problem with this migration is the dependency on `PortalSettings`. This demonstrates the tight-coupling that top-level code has with bottom-tier code and vice-versa which makes these changes very difficult.

My proposed solution is to create a new `IPortalSettings` in the `DotNetNuke.Abstractions` project. Then I updated the actual `PortalSettings` in `DotNetNuke.Library` to implement the class and updated code throughout. The reach of this change is pretty significant, be sure to take a close look at the following items:

* `IPortalController` - I updated the GetCurrentPortalSettings() to be deprecated in lieu of changing the signature. If we were to change the signature to return `IPortalSettings` it will create a ripple effect and that is a much larger change. There is now a new method on the interface called `GetCurrentSettings()` which returns the `IPortalSettings`
* Updated several usages of `PortalSettings` to `IPortalSettings` - I believe these changes are not breaking changes, even for public APIs, this is because `PortalSettings` now inherits from `IPortalSettings`
* I added IPortalController as a Dependency Injection registration so we can easily resolve it in the NavigationManager. This is the first instance Dependency Injection in the platform code where we have a dependency resolving another dependency. It is finally paying off 🎉

ahoefling commented:
I fixed the failing unit tests. I needed to create a mock of `IServiceProvider` and `INavigationManager` and manually assign it to the `Globals.DependencyProvider`. The difficult part to this is handling the state of Singletons. DNN uses the Singleton pattern throughout the codebase, the usage of this pattern was caching state between test classes which demonstrates poor unit test isolation. 

As we move forward, hopefully we can migrate these tests to use interfaces instead of the singleton pattern

ahoefling commented:
Running `./build.ps1 -Target CreateInstall` isn't working and I am waiting for build changes to be finalized from #3137. I have custom built an installer for this Pull Request but our installer files are too large to upload to GitHub. I have ran through the installer I have and this change does install correctly and basic DNN functionality appears to work.

If anyone wants to help verify this change I can email you the installer file or share it in one of the many slack channels we have in the DNN Community.

mitchelsellers commented on Build/Tools/NuGet/DotNetNuke.Abstractions.nuspec:
```diff
@@ -0,0 +1,22 @@
+    <description>
+      Provides common abstraction API references for the DNN Platform.
+    </description>
+    <copyright>DNN and DotNetNuke are copyright 2002-2019 by DNN Corp. All Rights Reserved.</copyright>
```
This should be adjusted. The proper copyright going forward for all items is

Copyright (c) .NET Foundation and Contributors

ahoefling replied:
Perfect! Thanks for the clarification on this

ahoefling replied:
@mitchelsellers should we create a task to update all file headers?

ahoefling replied:
I pushed changes that fixes this and updated all `.nuspec` files to use the correct copyright

ahoefling commented:
Since this is a large Pull Request, I saw it best to rebase on the latest changes. This is a common practice I have done on large PRs for other .NET open source projects (CoreFX).

The build updates appear to be working and I can now build this solution locally and get an installer that works for testing. Once the build completes check out the installer and see if it works for you as well. 

I would love to find problems with the changes now so we can address them during the code review

ahoefling commented:
@mitchelsellers @bdukes 
I have pushed some more changes to address some of our requested changes. Right now this PR is just about ready to merge. If you guys can review again and provide any additional feedback.

My next steps at this point:
* Test the installer again (earlier tests proved to be successful)
* Clean up PR to exclude extra content in `.config` and `.csproj`

ahoefling commented:
@bdukes @mitchelsellers 
I have tested the installer and marked all tasks completed. I believe this PR is ready to be merged.

Unless you have additional feedback
